File: Criacionais/Builder/Exemplo/builder.py
# Define as Partes de um Carro:
class Motor:
    def __init__(self, tipo):
        self.tipo = tipo
 
 
class Carroceria:
    def __init__(self, estilo):
        self.estilo = estilo
 
 
class Rodas:
    def __init__(self, tamanho):
        self.tamanho = tamanho
 
 
class Interior:
    def __init__(self, cor):
        self.cor = cor
 
 
# Builder
class CarroBuilder:
    def __init__(self):
        self.motor = None
        self.carroceria = None
        self.rodas = None
        self.interior = None
 
    def addMotor(self, tipo):
        self.motor = Motor(tipo)
        return self
 
    def addCarroceria(self, estilo):
        self.carroceria = Carroceria(estilo)
        return self
 
    def addRodas(self, tamanho):
        self.rodas = Rodas(tamanho)
        return self
 
    def addInterior(self, cor):
        self.interior = Interior(cor)
        return self
 
    def construir(self):
        return Carro(
            self.motor,
            self.carroceria,
            self.rodas,
            self.interior
        )
 
 
# Construir Carro
class Carro:
    def __init__(self, motor, carroceria, rodas, interior):
        self.motor = motor
        self.carroceria = carroceria
        self.rodas = rodas
        self.interior = interior
 
    def mostrarDetalhes(self):
        print("Carro composto por:")
        print("Motor:", self.motor.tipo)
        print("Carroceria:", self.carroceria.estilo)
        print("Rodas:", self.rodas.tamanho)
        print("Interior:", self.interior.cor)
 
 
# Cliente - Usando o Builder para construir o carro
builder = CarroBuilder()
carro1 = builder \
    .addMotor("1.4") \
    .addCarroceria("Esportiva") \
    .addRodas(18) \
    .addInterior("Preto") \
    .construir()
 
carro1.mostrarDetalhes()

File: Criacionais/Builder/Exemplo/builder.js
// Define as Partes de um Carro:
class Motor {
    constructor(tipo) {
        this.tipo = tipo
    }
}
 
class Carroceria {
    constructor(estilo) {
        this.estilo = estilo
    }
}
 
class Rodas {
    constructor(tamanho) {
        this.tamanho = tamanho
    }
}
 
class Interior {
    constructor(cor) {
        this.cor = cor
    }
}
 
 
 
// Builder
class CarroBuilder {
    constructor() {
        this.motor = null
        this.carroceria = null
        this.rodas = null
        this.interior = null
    }
 
    addMotor(tipo) {
        this.motor = new Motor(tipo)
        return this
    }
 
    addCarroceria(estilo) {
        this.carroceria = new Carroceria(estilo)
        return this
    }
 
    addRodas(tamanho) {
        this.rodas = new Rodas(tamanho)
        return this
    }
 
    addInterior(cor) {
        this.interior = new Interior(cor)
        return this
    }
 
    construir() {
        return new Carro(
            this.motor,
            this.carroceria,
            this.rodas,
            this.interior
        )
    }
}
 
// Construir Carro
class Carro {
    constructor(motor, carroceria, rodas, interior) {
        this.motor = motor
        this.carroceria = carroceria
        this.rodas = rodas
        this.interior = interior
    }
 
    mostrarDetalhes() {
        console.log("Carro composto por:")
        console.log("Motor:", this.motor.tipo)
        console.log("Carroceria:", this.carroceria.estilo)
        console.log("Rodas:", this.rodas.tamanho)
        console.log("Interior:", this.interior.cor)
    }
}
 
// Cliente - Usando o Builder para construir o carro
const builder = new CarroBuilder();
const carro1 = builder
    .addMotor("1.4")
    .addCarroceria("Esportiva")
    .addRodas(18)
    .addInterior("Preto")
    .construir();
 
carro1.mostrarDetalhes();